doc/tools/docgram/README.md

```diff
@@ -0,0 +1,194 @@
+empty production on the right-hand side.
+
+Another example: `LEFTQMARK: [ | "?" ]` is a local edit that treats `LEFTQMARK` as
+the name of a non-terminal and adds one production for it.  (We know that LEFTQMARK
+is a token but doc_gram does not.)  `SPLICE: [ | LEFTQMARK ]` requests replacing all
+uses of `LEFTQMARK` anywhere in the grammar with its productions and removing the
+non-terminal.  The combined effect of these two is to replace all uses of
+`LEFTQMARK` to `"?"`.
+
+Here are the current operations.  They are likely to be refined as we learn
+what operations are most useful while we update the mlg files and documentation:
+
+Global edits
+------------
+
+DELETE - deletes the specified non-terminals anywhere in the grammar.
+Useful for removing non-terminals that only do lookahead that shouldn't be
+in the documentation.
+
+RENAME - each production specifies an (old name, new name) pair of
+non-terminals to rename.
+
+SPLICE - requests replacing all uses of the nonterminals anywhere in the
+grammar with its productions and removing the non-terminal.  (Doesn't work
+recursively; splicing for both `A: [ | B ]` and `B: [ | C ]` has to be
+done in 2 SPLICE operations.)
+
+EXPAND_LISTS - expands LIST0, LIST1, LIST* ... SEP and OPT constructs into
+new non-terminals
+
+Local edits
+-----------
+
+DELETE <production> - removes the specified production from the grammar
+
+EDIT <production> - modifies the specified production using the following tags
+that appear in the specified production:
+  USE_NT <name> LIST* - extracts LIST* as new nonterminal with the specified
+    new non-terminal name
+  ADD_OPT <grammar symbol> - looks for a production that matches the specified
+    production **without** <grammar_symbol>.  If found, both productions are
+    replaced with single production with `OPT <grammar_symbol>`
+
+The current version handles a single USE_NT or ADD_OPT per EDIT.
+
+REPLACE <oldprod>
+WITH <newprod> (2 sequential productions) - removes <oldprod> and
+inserts <newprod> in its place.
+
+(any other NT name) - adds a new production (and possibly a new nonterminal)
+to the grammar.
+
+.rst file updates
+-----------------
+
+doc_gram updates the .rst file when it sees the following 3 lines
+
+```
+.. insertgram <start> <end>
+
+.. productionlist:: XXX
+```
+
+The end of the existing `productionlist` is recognized by a blank line or a line
+with `:name:` or `:undocumented:`
+
+Other details
+-------------
+
+The output identifies productions from plugins that aren't automatically loaded with
+`(* XXX plugin *)` in grammar files and with `     (XXX plugin)` in productionlists.
+If desired, this mechanism could be extended to tag certain productions as deprecated,
+perhaps in conjunction with a coqpp change.
+
+My impression is that the order of the productions affects parsing.  If this is
+correct, we should consider how appropriate it is to document the grammar as if
+it is a context free grammar.
```

Comment:
IINM this was confirmed, so it should be reformulated to keep only the last interrogation. My personal feeling is that documenting the grammar as a context free grammar is already much better than nothing.